fuel-core/src/schema/balance.rs

```diff
@@ -0,0 +1,192 @@
+        self.amount.into()
+    }
+
+    async fn asset_id(&self) -> AssetId {
+        self.asset_id.into()
+    }
+}
+
+#[derive(InputObject)]
+struct BalanceFilterInput {
+    /// Filter coins based on the `owner` field
+    owner: Address,
+}
+
+#[derive(Default)]
+pub struct BalanceQuery;
+
+#[Object]
+impl BalanceQuery {
+    async fn balance(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "address of the owner")] owner: Address,
+        #[graphql(desc = "asset_id of the coin")] asset_id: AssetId,
+    ) -> async_graphql::Result<Balance> {
+        let db = ctx.data_unchecked::<Database>();
+
+        let coin_ids: Vec<fuel_tx::UtxoId> =
+            db.owned_coins(owner.into(), None, None).try_collect()?;
+        let mut coins: Vec<(fuel_tx::UtxoId, CoinModel)> = coin_ids
+            .into_iter()
+            .map(|id| {
+                Storage::<fuel_tx::UtxoId, CoinModel>::get(db, &id)
+                    .transpose()
+                    .ok_or(KvStoreError::NotFound)?
+                    .map(|coin| (id, coin.into_owned()))
+            })
+            .try_collect()?;
+        coins.retain(|(_, coin)| coin.status == CoinStatus::Unspent);
+
+        let sum_amount = coins
+            .iter()
+            .filter(|(_, coin)| coin.asset_id == asset_id.into())
+            .map(|(_, coin)| coin.amount)
+            .sum();
+
+        Ok(Balance {
+            owner: owner.into(),
+            amount: sum_amount,
+            asset_id: asset_id.into(),
+        })
+    }
+
+    async fn balances(
+        &self,
+        ctx: &Context<'_>,
+        filter: BalanceFilterInput,
+        first: Option<i32>,
+        after: Option<String>,
+        last: Option<i32>,
+        before: Option<String>,
+    ) -> async_graphql::Result<Connection<AssetId, Balance, EmptyFields, EmptyFields>> {
+        let db = ctx.data_unchecked::<Database>();
+
+        let coin_ids: Vec<fuel_tx::UtxoId> = db
+            .owned_coins(filter.owner.into(), None, None)
+            .try_collect()?;
+        let mut coins: Vec<(fuel_tx::UtxoId, CoinModel)> = coin_ids
+            .into_iter()
+            .map(|id| {
+                Storage::<fuel_tx::UtxoId, CoinModel>::get(db, &id)
+                    .transpose()
+                    .ok_or(KvStoreError::NotFound)?
+                    .map(|coin| (id, coin.into_owned()))
+            })
+            .try_collect()?;
+        coins.retain(|(_, coin)| coin.status == CoinStatus::Unspent);
+
+        let balances = coins
+            .iter()
+            .group_by(|(_, coin)| coin.asset_id)
+            .into_iter()
+            .map(|(asset_id, coins)| {
+                let sum_amount = coins.map(|(_, coin)| coin.amount).sum();
+
+                Balance {
+                    owner: filter.owner.into(),
+                    amount: sum_amount,
+                    asset_id,
+                }
+            })
+            .collect::<Vec<Balance>>();
```

Comment:
Same thing here, we should be able to output these sums without collecting all the coins into a vec first.